### ISSUE TITLE
Notification alert in iOS8 causes Appium session start to fail due to reported instruments crash on startup

### ISSUE THREAD

**Issue by chrigrahcisco** [appium/appium#4178]
THE FIX / WORKAROUND IS AS FOLLOWS:

What you need to do is insert a 5-8 second sleep IN THE SOURCE CODE OF YOUR APP just before the notification code IN THE SOURCE CODE OF YOUR APP.  Caps are simply for emphasis so it's totally clear where the fix / workaround occurs as it has absolutely nothing to do with Appium. :)  

For example, I open my project in Xcode and in my app's source code I find the notification code and add this simple sleep line of code:

```
 // Register for notifications
 sleep(8);
```

Ask your developers to help you with this. I have the the ability to pull source and generate my own builds but still needed help from my dev team to locate the proper module and proper notification code to properly inject the sleep.

This will let Appium load your app. You can then handle the notification UIAAlert via your test code.
---
## Original problem report follows

Mac OSX 10.9.5
Xcode 6.1
Appium.app 1.3.3
Python 2.7

My AUT includes notification support. When installing the app on a real device on any iOS version you receive a startup notification dialog to allow / don't allow the app to send notifications. Using an iOS7.x simulator, this notification dialog is not presented. Based on initial research it appears a change with the latest Xcode and iOS8.x simulators allows this notification dialog to be presented within the simulator even though you cannot physically register the simulator for notifications.

This appears to present a problem when Appium initially tries to create the session, apparently because the AUT isn't successfully detected due to the notification, at least that is my guess at this time. I would say 90%+ of the time tests fail due to this issue, however occasionally it does manage to work and the tests proceed successfully.

Here is a snippet from the log.

2014-11-26 17:19:48:789 - info: [IOS_SYSLOG_ROW ] Nov 26 12:19:48 CHRIGRAH-M-2038 ScriptAgent[35039]: Warning - Target app is not frontmost

2014-11-26 17:19:48:815 - info: [debug] [INST] 2014-11-26 17:19:48 +0000 Fail: Could not start script, target application is not frontmost.

2014-11-26 17:19:48:925 - info: [IOS_SYSLOG_ROW ] Nov 26 12:19:48 CHRIGRAH-M-2038 com.apple.CoreSimulator.SimDevice.9C0D7887-CCD5-4A1A-BEBA-E079FBE1663E.launchd_sim[34839](UIKitApplication:com.MYAPP[0x6ac2][34981]): Service exited due to signal: Killed: 9

2014-11-26 17:19:49:036 - info: [IOS_SYSLOG_ROW ] Nov 26 12:19:49 CHRIGRAH-M-2038 SpringBoard[34850]: Application 'UIKitApplication:com.MYAPP[0x6ac2]' exited abnormally via signal.

2014-11-26 17:19:49:037 - info: [IOS_SYSLOG_ROW ] Nov 26 12:19:49 CHRIGRAH-M-2038 assertiond[34854]: notify_suspend_pid() failed with error 7
2014-11-26 17:19:49:037 - info: [IOS_SYSLOG_ROW ] Nov 26 12:19:49 CHRIGRAH-M-2038 assertiond[34854]: assertion failed: 13F34 12B411: assertiond + 11523 [3F572A0B-7E12-378D-AFEE-EA491BAF2C36]: 0x1

2014-11-26 17:19:49:698 - info: [debug] [INST] Instruments Trace Complete (Duration : 36.580814s; Output : /tmp/appium-instruments/instrumentscli0.trace)

2014-11-26 17:19:49:720 - info: [debug] [INSTSERVER] Instruments exited with code 0

2014-11-26 17:19:49:720 - info: [debug] Killall instruments

2014-11-26 17:19:49:722 - info: [debug] Instruments crashed on startup
2014-11-26 17:19:49:722 - info: [debug] Attempting to retry launching instruments, this is retry #1

**Comment by bthuan** [comment]
I have the same problem too: when I run the tests one by one it is ok. But when I run them as a bunch ( using maven) it beaks all the time.
Hopefully they will fix it with 1.3.5

**Comment by chrigrahcisco** [comment]
I installed 1.3.5-beta version of Appium.app using this link: https://bitbucket.org/appium/appium.app/downloads/appium-1.3.5-beta.dmg

I added this to my desired_capabilities: desired_caps['waitForAppScript'] =  True

However, I'm not seeing any type of logging that indicates this is being utilized.

Unfortunately, the issue does not appear to be corrected and my tests continue to fail to execute a significant amount of the time (86% fail rate on my last test run). Let me know what I can do to help expedite a successful resolution.


PERTINENT APPIUM LOG SNIPPET
--------------------------------------------
2015-01-27 19:24:38:441 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 ScriptAgent[15360]: Warning - Target app is not frontmost

2015-01-27 19:24:38:479 - info: [debug] [INST] 2015-01-27 19:24:38 +0000 Fail: Could not start script, target application is not frontmost.

2015-01-27 19:24:38:588 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 SpringBoard[15207]: Unable to get short BSD proc info for 15286: No such process

2015-01-27 19:24:38:596 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 com.apple.CoreSimulator.SimDevice.FBDF8D36-411C-46A6-A39D-3C609C79C079.launchd_sim[15196] (UIKitApplication:MYAUT[0xc789][15286]): Service exited due to signal: Killed: 9

2015-01-27 19:24:38:705 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 SpringBoard[15207]: Application 'UIKitApplication:MYUAT[0xc789]' exited abnormally via signal.
2015-01-27 19:24:38:705 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 assertiond[15211]: assertion failed: 14B25 12B411: assertiond + 11523 [3F572A0B-7E12-378D-AFEE-EA491BAF2C36]: 0x1
2015-01-27 19:24:38:706 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 assertiond[15211]: notify_suspend_pid() failed with error 7
2015-01-27 19:24:38:706 - info: [IOS_SYSLOG_ROW ] Jan 27 14:24:38 CHRIGRAH-M-2038 assertiond[15211]: assertion failed: 14B25 12B411: assertiond + 11523 [3F572A0B-7E12-378D-AFEE-EA491BAF2C36]: 0x1

2015-01-27 19:24:39:231 - info: [debug] [INST] Instruments Trace Complete (Duration : 36.381435s; Output : /tmp/appium-instruments/instrumentscli0.trace)

2015-01-27 19:24:39:254 - info: [debug] [INSTSERVER] Instruments exited with code 0
2015-01-27 19:24:39:254 - info: [debug] Killall instruments

2015-01-27 19:24:39:259 - info: [debug] Instruments crashed on startup

**Comment by jlipps** [comment]
@chrigrahcisco the `waitForAppScript` cap is not a boolean--you actually need to write in a uiautomation js script in a string that appium will execute until _it_ returns true

**Comment by hallbeck** [comment]
A piece of sample code would be very helpful. Since this impacts most users, it should be as simple as possible.

**Comment by jlipps** [comment]
@hallbeck I disagree, Appium's default script should be sufficient for most users. The `waitForAppScript` cap is available for people for whom the default script doesn't work, and who know what they are doing. Are you saying the default script isn't working for most people? The default script simply waits for there to be at least one element in the UI hierarchy.

If it turns out that's a bad default script, we should replace it with a more functional one. Not sure what that would be, though.

**Comment by hallbeck** [comment]
Im not doing anything tricky, or complicated. Our app doesn't either. But if I run my tests on a iOS8.x simulator I have to manually dismiss the "we would like to send you push notifications". On iOS 7.x it doesn't have the popups because iOS changed how they asked in iOS8. My tests were running fine until recently. Even when the pop up was there I could click OK in code.
I'll have to check when I wrote the first issue on it here, but that was when it all broke for me. I believe that many have this problem and their workaround is to test on iOS 7 for now till you guys fix it. As it stands I cannot have my tests kick off and run without me sitting here to click OK. I tried the dismiss alerts, and that's fine if I'm not testing that a popup comes up in an error condition. So, I stopped doing that.
I don't know about everyone else. But I need that example. I'm not "good" enough to be able to "just figure it out and write in a uiautomation js script ...." that completely lost me. I just want it to work.

**Comment by jlipps** [comment]
@hallbeck I agree that it would be ideal for appium to just work and handle this alert case. I'm not sure what that has to do with the `waitForAppScript` cap or why @sebv thinks that cap can handle this case. I guess we're talking about different things.

**Comment by chrigrahcisco** [comment]
I don't know what the under-the-hood root cause is. I only know that Appium has been basically unusable to me for 2 months due to the fact I am testing using the iOS8 simulator which always throws the Notification dialog each and every time a new test is executed.

I'm happy to try to provide whatever data you require, but you'll need to provide me with detailed instructions on how to obtain that data.

**Comment by hallbeck** [comment]
right, I thought that they had fixed it and we were just waiting for the fix to appear. it seems they have no idea what we are talking about. <sad>

**Comment by chrigrahcisco** [comment]
They say a picture is worth 1,000 words. Here is a screenshot of the dialog in question. What happens is the app launches, then this modal dialog pops up on top of the app, and apparently Appium then no longer can find the app since it's no longer focused.

![notification dialog ios8 simulator app launch](https://cloud.githubusercontent.com/assets/7501032/5927392/2db65568-a640-11e4-9ff3-762535683087.png)

**Comment by jlipps** [comment]
It sounds to me like this is an Apple UIAutomation bug wherein when an app requests permissions on launch, a dialog launches that effectively gets slotted in before the UIAutomation context gets going, thus blocking any automation from occurring. This results in a catch-22: we need UIAutomation support for Appium to be able to handle this dialog, but we can't get UIAutomation support activated until this dialog is dismissed.

The only known workaround for this Apple bug is delaying (in the app code) the request for permissions by a few seconds (in older versions of iOS, people reported that about 8s was solid). This gives UIAutomation a chance to spin up, so that by the time the dialog appears, we have an automation context we can use to to handle the alert (using e.g., the `autoAcceptAlerts` capability).

There might be a longer-term solution here, on the analogy of the `locationServicesAuthorized` capability. Maybe there's a way for us to pre-populate the notifications permissions database, so that we could add a `notificationsAuthorized` capability that would mean the app starts with notifications already authorized, resulting in no dialog at all.

But either way, given that I'm jumping into this conversation with less context than I thought (reading back a little bit, I thought it was about the `waitForAppScript` cap, when it's really about automation being blocked by an alert on app load). So I'm curious why @sebv thinks this problem might be solvable in conjunction with `waitForAppScript`.

I hope this helps---is it possible for you to try the workaround to see if it at least unblocks you for the time being?

Another thing you could try is setting the `waitForAppScript` capability to the _string_ `'true'`, or the string `'$.delay(5000); true'`. This could help if it's Appium itself that is confused because it doesn't think the app is loaded---but from the logs it doesn't look like this is the case.

The long long term solution is of course for Apple to fix their bug. To that end, what would be helpful is for you to try to reproduce this using UIAutomation alone (not Appium), to show that it's not possible to use their automation framework when an app that requests permissions on launch.

**Comment by hallbeck** [comment]
the thing is it was working fine in 1.3.1 (I think). then I stupidly upgraded and it all broke. I can't go back now because 1.3.1 can't run on 8.1 (only 8.0 I think) and my Xcode won't build for 8.0. my co-worker did not have the problem while running the latest Xcode with 1.3.1 server. then she upgraded and has the same problem. So, what do the masses do? I don't have time to debug appium vs Xcode. there are whole companies to do that. I just need my tests to run.

**Comment by gempesaw** [comment]
Is there a way I can set an Appium test up as a REPL to arbitrary UIAutomation JS to try to figure out how to dismiss the notification alert? I'd like to try a couple different scripts out and not have to restart the entire Appium session between different iterations of the waitForAppScript.

Or, is the best way to do that to just Instrument the app in XCode and try it out in there (this is probably a fine way to do it)? If so, my followup question is what is the `$.`,  that I see popping up in Appium's UIAutomation JS, does it come from `bootstrap.js`? 

Thaaaanks :P

**Comment by gempesaw** [comment]
@hallbeck @chrigrahcisco Setting waitForAppScript to 

```js
$.delay(3000); $.acceptAlert();
```

worked for me. You'll probably have to tinker with the delay time on your own. The $.acceptAlert() is just what autoAcceptAlerts calls on its own anyway from uiauto, except we're not setting the onAlert handler, we're only handling this first alert. 

relevant parts of the log:

```
info: --> POST /wd/hub/session {"desiredCapabilities":{"platformName":"iOS","app":"REDACTED","waitForAppScript":"$.delay(3000); $.acceptAlert();","sendKeyStrategy":"grouped","newCommandTimeout":999999,"deviceName":"iPhone 6","platformVersion":"8.2"}}
```

```
info: [debug] Using custom script to wait for app start:$.delay(3000); $.acceptAlert();
info: [debug] Pushing command to appium work queue: "try{\n$.delay(3000); $.acceptAlert();\n} catch(err) { $.log(\"waitForAppScript err: \" + error); false; };"
info: [debug] Sending command to instruments: try{
$.delay(3000); $.acceptAlert();
} catch(err) { $.log("waitForAppScript err: " + error); false; };
info: [debug] [INST] 2015-01-28 01:09:14 +0000 Debug: Got new command 2 from instruments: try{
       $.delay(3000); $.acceptAlert();
} catch(err) { $.log("waitForAppScript err: " + error); false; };
info: [debug] [INST] 2015-01-28 01:09:14 +0000 Debug: evaluating try{
       $.delay(3000); $.acceptAlert();
} catch(err) { $.log("waitForAppScript err: " + error); false; };
info: [debug] [INST] 2015-01-28 01:09:17 +0000 Debug: target.frontMostApp().alert().buttons()[1].tap()
info: [debug] [INST] 2015-01-28 01:09:17 +0000 Debug: Waiting for alert to close...
info: [debug] [INST] 2015-01-28 01:09:17 +0000 Debug: Waiting for alert to close...
info: [debug] [INST] 2015-01-28 01:09:18 +0000 Debug: evaluation finished
info: [debug] [INST] 2015-01-28 01:09:18 +0000 Debug: responding with:
info: [debug] [INST] 2015-01-28 01:09:18 +0000 Debug: Running system command #3: /usr/local/Cellar/node/0.10.28/bin/node /Users/dgempesaw/opt/jlipps-appium/node_modules/appium-uiauto/bin/command-proxy-client.js /tmp/instruments_sock 2,{"status":0,"value":""}...
```

and a totally useless-to-everyone-else perl version of my script, haha :D

```
#! /usr/bin/perl

use strict;
use warnings;
use Appium;

my $caps = {
    app => 'REDACTED',
    waitForAppScript => '$.delay(5000); $.acceptAlert();',

    deviceName => 'iPhone 6',
    platformName => 'iOS',
    platformVersion => '8.2',
    newCommandTimeout => 999999,
    sendKeyStrategy => 'grouped',
};

my $appium = Appium->new(caps => $caps);

print $appium->find_element('Login', 'name')->get_text; # prints "Login" without me having to handle the alert

$appium->quit;
```

Oddly, it ran the delay/acceptAlert a few times. But, it didn't throw any errors, it cleared my alert, and I was able to go about my merry way with my automation. 

Does it work for you ?

**Comment by chrigrahcisco** [comment]
I'm unable to make this work on my end. I do see the waitForAppScript listed on the desiredCapabilities section of the Appium log. I do see it successfully closing the Notification alert in some, but not all, cases. 

I've bumped the delay(5000) to delay(10000) and then delay(30000) assuming a longer delay will theoretically increase the chance of it successfully closing the alert but it appears to make no real difference in the end result.

I don't have time right now to dig deeper at this point.

'''
2015-01-28 15:04:00:734 - info: [debug] [INST] 2015-01-28 15:04:00 +0000 Debug: target.frontMostApp().alert().buttons()[1].tap()
2015-01-28 15:04:00:874 - info: [debug] [INST] 2015-01-28 15:04:00 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:01:153 - info: [debug] [INST] 2015-01-28 15:04:01 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:01:566 - info: [debug] [INST] 2015-01-28 15:04:01 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:01:857 - info: [debug] [INST] 2015-01-28 15:04:01 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:02:125 - info: [debug] [INST] 2015-01-28 15:04:02 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:02:408 - info: [debug] [INST] 2015-01-28 15:04:02 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:02:684 - info: [debug] [INST] 2015-01-28 15:04:02 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:03:101 - info: [debug] [INST] 2015-01-28 15:04:02 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:03:381 - info: [debug] [INST] 2015-01-28 15:04:03 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:03:659 - info: [debug] [INST] 2015-01-28 15:04:03 +0000 Debug: Waiting for alert to close...
2015-01-28 15:04:03:936 - info: [debug] [INST] 2015-01-28 15:04:03 +0000 Debug: Waited for a while and alert didn't close, moving on
'''

**Comment by B-Dubs** [comment]
I'm seeing the same thing.  When the app starts up, it sits at this point in the Appium log...

info: [debug] [INST] Waiting for device to boot...

...Until it either times out and fails, or until I manually dismiss the notifications alert popup.  For now, I have been starting appium with the --no-reset flag so the alert doesn't have to launch.

I like @jlipps suggestion of adding a "notificationsAuthorized" capability that would mean the app starts with notifications already authorized, resulting in no dialog at all.

**Comment by hallbeck** [comment]
the thing is.. we use this for testing, right? there should be some way to test that. if we automatically dismiss it, we can't test it. But I'll take it is that's all we can get. We can manually test that.

**Comment by chrigrahcisco** [comment]
To my knowledge the Notifications alert has no functional behavior on the simulator other than being invoked. You can click OK all you want, it doesn't actually register the simulator with notification services. So the only testing you could possibly do is is to validate that it's physically appeared. 

I do all of my automation on simulators at this time, so I cannot comment on how Appium behaves against a real piece of hardware in regards to handling the session when the Notification dialog appears.

------------------

Back to waitForAppScript...

Does anyone have any ideas on the root cause as to why the script typically fails to successfully close the Notification alert? A failure forces it to retry, which may or may not be successful. I've bumped my backend retries to 10 and have only experienced 2 "Instruments crashed on startup" failures in 17 test case executions thus far, that same thing had no positive effects using 1.3.4   

My current settings:

desired_caps['waitForAppScript'] =  '$.delay(30000); $.acceptAlert();

Server Flags: --backend-retries 10

This is a bandaid though, the massive retries causes a serious slowdown in the tests actually executing. But at least it's progress towards them executing eventually.

I'll likely have time on Friday to dig deeper if someone provides some detailed directions on what to try and what to look for.

**Comment by gempesaw** [comment]
Curiously, this works when my Appium server is on Yosemite, but not when my Appium server is on Mavericks. Not sure if that's reproduce-able, or why it's happening, but. On mavericks, my tests are super flaky and crashy; on yosemite they are rock solid.

**Comment by mbknieri** [comment]
I am having the same issues as @chrigrahcisco with the suggested solution made by @gempesaw. Sometimes the waitForAppScript will execute, the alert will get dismissed, and the automation continues as normal. Sometimes, it seems that Appium hangs on waiting for the device to boot and instruments will crash:

info: [debug] And launch timeouts (in ms): {"global":90000}
info: [debug] [INST] Waiting for device to boot...
info: [debug] [INST] 2015-01-29 16:22:51 +0000 Fail: Could not start script, target application is not frontmost.
info: [debug] [INST] Instruments Trace Complete (Duration : 46.821762s; Output : /tmp/appium-instruments/instrumentscli0.trace)
info: [debug] [INSTSERVER] Instruments exited with code 0
info: [debug] Killall instruments
info: [debug] Instruments crashed on startup
info: [debug] Attempting to retry launching instruments, this is retry #1
info: [debug] Killall iOS Simulator
info: Launching instruments
info: [debug] Attempting to run app on iPhone 6 (8.1 Simulator)

This is with waitForAppScript=$.delay(5000); $.acceptAlert();

I am running on Mavericks.

**Comment by bthuan** [comment]
@gempesaw I don't know if something wrong with my machine but I am running Appium on Yosemite with waitForAppScript cap and my tests are still unstable. Sometimes the alert is dismissed, but sometimes it is not and Appium hangs on waiting for the device to reboot ( same problem as @mbknieri ) . In this case, it makes the following tests failed which is so bad

**Comment by gempesaw** [comment]
Yeah, I ran a larger suite of tests and found that the waitForAppScript cap is flaky with the exact problems listed above `Waiting for device to boot...`. Apparently the cap is supposed to return true when it's done, so I've been using `$.delay(5000); $.acceptAlert(); true;` but with the same inconsistent results. I'm back to `autoAcceptAlerts: true` for now. Sad panda, thought we had it licked ._.

**Comment by mbknieri** [comment]
I'm still facing the same issue while setting autoAcceptAlerts: true.

I vote up for exploring this idea proposed by @jlipps:

>Maybe there's a way for us to pre-populate the notifications permissions database, so that we could add >a notificationsAuthorized capability that would mean the app starts with notifications already authorized, >resulting in no dialog at all.

**Comment by chrigrahcisco** [comment]
Any update on this issue?   I haven't been able to use Appium effectively since November 2014.

**Comment by jlipps** [comment]
@chrigrahcisco did you try the workaround of delaying the permissions check in your app code by a few seconds? i know it's not a good long-term solution, but it will tell us if what we're dealing with here is indeed the apple bug.

**Comment by mbknieri** [comment]
I'm testing the latest release now (v1.3.5). I've passed "$.delay(5000); $.acceptAlert(); true;" into the waitForAppScript capability. Sometimes instruments crashes with "Fail: Could not start script, target application is not frontmost." Sometimes the waitForAppScript kicks in and dismisses the alert. Without passing in this value to the waitForAppScript capability, Instruments crashes with the "application is not frontmost" error.

**Comment by chrigrahcisco** [comment]
@jlipps I somehow missed the suggestion to attempt delaying permissions. I'll request assistance from my dev team on exploring that and report back.

Note that autoAcceptAlerts is not a viable option for my project as the AUT leverages a lot of UIAAlert dialogs throughout the app and I wouldn't want them all simply dismissed automatically. But I do already have successful code that can recognize the Notification alert and dismiss it for me, so as soon as Appium can reliably and consistently start a session I'm good to go.

**Comment by chrigrahcisco** [comment]
@jlipps A simple    `sleep(8);`  before the block of Notification code appears to work fine. Appium session appears to reliably start, and my startup code can detect and close the Notification dialog successfully.  I can also simply comment out the block of Notification code to turn that off completely.

**Comment by Demappa** [comment]
Hi Guys,

Even I have Same Problem.I Need to click on an alert and this alert will popup as soon as application is launched  it was working fine with 1.3.3 but when i updated Appium to 1.3.4 I'm not able to click on alert I'm getting below error

"" Could not start script, target application is not frontmost.
info: [debug] [INST] Instruments Trace Complete (Duration : 33.088829s; Output : /tmp/appium-instruments/instrumentscli0.trace)"

Please help me I'm badly sucked with this problem

**Comment by gempesaw** [comment]
> A simple sleep(8); before the block of Notification code appears to work fine. Appium session appears to reliably start, and my startup code can detect and close the Notification dialog successfully. I can also simply comment out the block of Notification code to turn that off completely.

@chrigrahcisco Are you talking about a sleep in your AUT or in your test ? :)

Also, against 1.3.5 I'm currently using `waitForAppScript: "$.delay(3000); UIATarget.localTarget().frontMostApp().alert(); $.acceptAlert(); true"`. Even though `$.acceptAlert()` is making doing exactly the same `target.frontMostApp.alert`, I've had pretty good success with this. Sometimes the appium session won't catch it on the first load of the simulator, but when appium restarts the simulator and tries attempt \#2, it dismisses the alert.

**Comment by chrigrahcisco** [comment]
To clarify:

* `Sleep(8);` in the actual source code of my AUT, added before the code that executes the Notification functions. This allows Appium to find the AUT so your tests will function. You will have to ask your dev team where this block of code is located.
* Or completely comment out the Notification functions in the source code, which will prevent the Notification alert from being invoked at all. If you're testing on the sim, Notifications are non-functional so this isn't an issue for me, personally.
* I'm not using `autoAcceptAlerts`  nor  `waitForAppScript`   Neither of those proposed solutions were effective. 

I'm not proposing this is the real / desired fix for the issue. But it is a viable workaround so you can get on with your testing for now.

**Comment by SenhorCastor** [comment]
Running Appium 1.3.6, still having this problem

**Comment by khallbec** [comment]
I can't believe no one is assigned to this. Its been broken for more than 3 months.

**Comment by jlipps** [comment]
@khallbec please read the previous comments; this is an Apple bug, not an Appium bug

**Comment by qazimusab** [comment]
Anyone know if 6.3 released a fix for this?

**Comment by SConrad-Whoop** [comment]
@chrigrahcisco Could you please post the full workaround snippet you are using? Some of us that are newer to Appium are a little lost. Maybe it's just me. I understand you're using the sleep, but I've yet to find code that actually allows the popup to be dismissed.

**Comment by chrigrahcisco** [comment]
Unfortunately this isn't something you do in Appium. It's something you'll need to do directly in the app source code using Xcode prior to building the app for testing. I cannot share my app source code, though the notification section doesn't seem particularly special or proprietary, but rules are rules. :)

Go talk to your development team and ask them how to find the Notification code. You may be able to find it yourself by searching for `UIUserNotificationSettings` Once that's located just stick `sleep(5);`  before the block of code that invokes the notification setup for iOS8 and then create your app for testing and that should do the trick.

**Comment by khallbec** [comment]
Its a bummer that its an apple bug. how do other automation tools get around it? or it would make most automation not unable. most companies are unwilling to add in special code for testers. I just sit and babysit my tests which make them far from automated. How does anyone run their stuff on sauce labs?

**Comment by LiudasTumasonis** [comment]
Every single time the app launches I try to accept the Notifications alert and it works only once in a while... And I just ran out of ideas on how to take care of this. Tried delays, multiple accept commands etc. Anyone has any updates on this?

**Comment by chrigrahcisco** [comment]
I've done my best to describe the successful workaround for this issue of inserting a sleep in the source code just before the notification code gets called. If you cannot figure out how to insert the sleep into the source code yourself please ask your dev team to help you.

**Comment by SConrad-Whoop** [comment]
@chrigrahcisco Ah, I misunderstood, thanks for clarifying.

**Comment by sebv** [comment]
Summary: Apple bug.

**Comment by jlipps** [comment]
we should list any apple bugs at https://github.com/appium/worms (after submitting a radar). follow the pattern of the current issue

**Comment by popei69** [comment]
Not sure if someone still get this issue.

For me it works fine with waitForAppScript and alertAccept (in JAVA):

Setting first my capabilities
`capabilities.setCapability("waitForAppScript", "$.delay(1000);");`

Calling alertAccept just after the app was launched and after the delay.
`driver.switchTo().alert().accept();`

**Comment by igalep** [comment]
@popei69 , Thanks for the input , but it doesn't work for me ...
I'm setting up the capabilities as suggested , and then I'm creating the driver and calling the .switchTo().....

on the first application launch it doesn't work , seems that the notifications pops up before the instrument were actually launched , after few seconds appium tries to launch the app once again , during the second time instruments launched before the pop up has shown and it's works then.

The question is how to make it work on the first attempt ?

**Comment by jdougit** [comment]
got the exact the same issue described here...sigh
tried wait script, no luck, "could not start script, target application is not frontmost"...
@igalep, are you on ios or android?  how to remove the registry ?

**Comment by igalep** [comment]
@jdougit I'm running on ios simulator 8.4 ver

you are right , appium restart the simulator and it does work for the second time ,but as I said , I just added a delay for the notification registration within the application code

**Comment by alangithubtrader** [comment]
@gempesaw  Thanks ! This works for me.

Setting waitForAppScript to

$.delay(3000); $.acceptAlert();

**Comment by oakdale2008** [comment]
@jlipps I am still having this problem in ios 9.0 with appium 1.4.11

I tried waitForAppScript = $.delay(3000); $.acceptAlert();
I also used  autoAcceptAlerts: true

However sometimes instruments crashes on startup due to this issue.

Any suggestions how to resolve welcome.

Thx!

**Comment by jtomson-mdsol** [comment]
We have a similar issue with notifications and appium, and we use a workaround that installs the app with notifications already accepted - maybe it might help others dealing with this?

The following snippets are in ruby but it's just shelling out to use `PlistBuddy` and `xcrun`

##### Start up the desired simulator and install the application

`SIM_DEVICE_NAME` is the readable name of the sim, 'e.g' iPhone
```
system "xcrun simctl boot '#{SIM_DEVICE_NAME}'"
system "xcrun simctl install '#{SIM_DEVICE_NAME}' '#{PATH_TO_APP}'"
```

##### Update the backboard plist to allow notifications
The target is a plist living at `$SIM_DEVICE_PATH/data/Library/BackBoard/applicationState.plist`
`APP_BUNDLE_ID` is your bundle identifier e.g. `com.foo.bar`
`SIM_DEVICE_PATH` is the path to the simulator, e.g. `~/Library/Developer/CoreSimulator/Devices/4D2D127A-7103-41B2-872B-2DB891B978A2`
```
# set s3kr1t plist entries to 'auto-accept' local & remote notifications
app_state_plist_path = File.join(SIM_DEVICE_PATH, "data", "Library", "BackBoard", "applicationState.plist")

buddy_cmds = ["add #{APP_BUNDLE_ID}                                                 dict",
              "add #{APP_BUNDLE_ID}:SBApplicationAllEnabledNotificationTypes        integer",
              "add #{APP_BUNDLE_ID}:SBApplicationAllSupportedNotificationTypes      integer",
              "add #{APP_BUNDLE_ID}:SBApplicationHaveRegisteredUserNotificationsKey bool",
              "add #{APP_BUNDLE_ID}:SBSPushStoreNotificationsDeniedByUser           bool",
              "set #{APP_BUNDLE_ID}:SBApplicationAllEnabledNotificationTypes        7",
              "set #{APP_BUNDLE_ID}:SBApplicationAllSupportedNotificationTypes      7",
              "set #{APP_BUNDLE_ID}:SBApplicationHaveRegisteredUserNotificationsKey true",
              "set #{APP_BUNDLE_ID}:SBSPushStoreNotificationsDeniedByUser           false"]

buddy_cmds.each do |cmd|
  puts "Executing PlistBuddy command: #{cmd}"
  system "/usr/libexec/PlistBuddy -c '#{cmd}' '#{app_state_plist_path}'"
end
```

**Comment by rams78** [comment]
iOS 9.0.1 Has weird issue - after an application is installed it pops up the 'Notifications' confirmation dialog - even though 'launch_app' has not been called yet. Needless to say it's completely blocking automation on real devices. Is anybody else experiencing same issue?

**Comment by ShankarGargXebia** [comment]
Setting waitForAppScript to $.delay(3000); $.acceptAlert(); worked for me
Appium GUI : 1.4.8 and language used: Java

**Comment by balajilinks** [comment]
Setting waitForAppScript to $.delay(3000); $.acceptAlert(); worked for me too :)
Appium GUI : 1.4.13 and language used: Java

**Comment by michaelgallacher** [comment]
@rams78 Yes. It's unfortunately a total timing thing.  My tests can dismiss the alert 95% of the time on my iPhone 5 and fail to dismiss the alert (thus aborting the test) 95% of the time on my iPhone 6S.

**Comment by RedPlanetTesting** [comment]
Facing the same issue with latest Appium as it is unable to identify the Notification alert and repeatedly launching the app under the uninfluenced Notification alert , Tried all the above suggested workarounds..
but handling  Notification alert seems to inconsistent. 
Please let me know if anyone found any updates on this issue. 
Appium should recognize this issue as a Sev 1 bug

**Comment by manishmittal79** [comment]
Same here. With 9.2 this has become worse. I tried so many different combinations inside the desired cap, but the driver is not able to switch to alert view. Please Appium devs, this is important, please provide any details to overcome this for time being. I am using Python.

**Comment by khallbec** [comment]
I WISH Appium would take this seriously. I cannot integrate with Jenkins because my tests never run through. I have to babysit them to catch it when it can't see that it has booted. it kind of defeats the purpose of automated tests. Its on my review to get them into Jenkins <sigh> Im about to put them in and have them all fail.

**Comment by chrigrahcisco** [comment]
I know this thread is very, very long with a lot of various suggestions and ideas tossed out. Let me recap by stating again that the fix / workaround has nothing to do with Appium and everything to do with your application's source code. I do not auto accept alerts as it doesn't affect this specific issue, and it literally closes every alert and my AUT uses alert dialogs all over the place for What's New, Logoff confirmation, etc.

What you need to do is insert a 5-8 second sleep IN THE SOURCE CODE OF YOUR APP just before the notification code IN THE SOURCE CODE OF YOUR APP.  Caps are simply for emphasis so it's totally clear where the fix / workaround occurs. :)  

For example, in my app's source code I find the notification code and add this simple sleep line of code:

     // Register for notifications
     sleep(8);

Ask your developers to help you with this if you do not have the ability to pull source and generate your own builds.

**Comment by manishmittal79** [comment]
Thanks for clarifying. I can try to ask the dev team to add the sleep. I understand that, It’s a kind of race condition for the driver to setup before the notification pops up.

**Comment by jlipps** [comment]
Hopefully moving to appium-xcuitest-driver in the future will free us from silly Apple bugs like this.

**Comment by ShankarGargXebia** [comment]
By when this change is expected?

~
Thanks
Shankar
+91-9873842038

Sent from mobile, excuse typos

On 16-Dec-2015, at 3:48 AM, Jonathan Lipps <notifications@github.com<mailto:notifications@github.com>> wrote:


Hopefully moving to appium-xcuitest-driver in the future will free us from silly Apple bugs like this.

-

**Comment by jlipps** [comment]
@ShankarGargXebia you can already run it from source https://github.com/appium/appium-xcuitest-driver. It has not yet been merged into Appium (will happen after 1.5 is released).

**Comment by islandev** [comment]
@popei69  thanks .
i worked  use what @popei69  do 

###   in desired_capabilities
```python
'waitForAppScript' :'$.delay(5000); $.acceptAlert();',
```

###  in test scrtipt
```python
self.driver.switch_to.alert.accept()
```

**Comment by alangithubtrader** [comment]
Hey guys, I think this is a more elegant way of handling this issue.

`UIATarget.onAlert = function onAlert(alert){return false;}`

**Comment by manishmittal79** [comment]
Is there a way to use this in python?
I am having trouble with dismissing the alerts using autoAcceptAlerts in
desired caps for iOS9.2.1.
For iOs9.2 and lower, it works fine but not with recent apple update.

Finally, I have to disable jenkins job as all tests were failing, as appium
driver died and instrument crashed after time out. The XML tree of the app
is also not available in the appium inspector due to this.

Regards
Manish

**Comment by cmmanish** [comment]
Thanks a Bunch, It works like a Charm

**Comment by chrigrahcisco** [comment]
I've had really mixed results attempting to use waitForAppScript, it often fails and requires the sim to be restarted again costing me more execution time. 

The best way to overcome this is still to inject a sleep(8); in your source code just before the notification code is invoked.

**Comment by almsta** [comment]
Same thing, injecting the `sleep(8)` works.

**Comment by chrigrahcisco** [comment]
It could be pure coincidence but right now I'm using Appium 1.5.1, Xcode 7.2 and iOS 9.2 Simulator and it seems to be working successfully without a sleep injected into the source code, nor any waitForAppScript attempt.

**Comment by nirajsingh1486** [comment]
This isn't fixed yet in Appium 1.4.13 and unfortunately i can't ask dev team to put sleep()  before popup.

Is there any solution yet or still this problem persist?

**Comment by chrisjywu** [comment]
same here, having problem with 1.4.13.  ios automation always fail to start test in the middle of the run. 

sticked with 1.4.13 because seems 1.4.13 is the newest version that supports android toast message check (in selendroid mode).

**Comment by chrisjywu** [comment]
But thanks I will try the inject 8 seconds workaround!

**Comment by chrisjywu** [comment]
@chrigrahcisco if you use sleep before the notification, does that mean that it will slow down the app launch time?

**Comment by chrigrahcisco** [comment]
@chrisjywu Correct, you're imposing a hard sleep directly in your app source code, that will slow down the app launch by however many seconds you're sleeping, I was using 8 seconds. However, your tests will now run reliably which is way more important than losing 8 seconds each time the app launches.

Note I no longer have to use the source code sleep, nor any 'waitForAppScript' desired_cap with with Appium 1.5.x   There appears to be excellent handling of waiting for the app at startup now. At least with the Xcode 7.2.1 iOS 9.x simulators I've personally tried.

**Comment by chrisjywu** [comment]
@chrigrahcisco thanks for your info!

**Comment by nirajsingh1486** [comment]
@chrigrahcisco Is this handled for real devices? I don't think so because we are badly stuck with this issue.any solution or still we in the same situation which we were one year back

**Comment by chrigrahcisco** [comment]
@nirajsingh1486   I have only ever tested iOS using simulators which seem to work perfectly fine without any source code sleep or other action on our part when using the latest Appium 1.5.x server. I cannot speak to hardware devices, though you can always try the hardcoded source code sleep.

**Comment by manishmittal79** [comment]
Hi,
Di you try using autoacceptalert in desired caps? If not try this one.
'autoAcceptAlerts': True

**Comment by sankritraj** [comment]
@manishmittal79 This has already been discussed and tried but not result if popup appears before app launch

**Comment by smaspe** [comment]
@jlipps  I see on the roadmap that the integration of xcuitext-driver is planned for 1.6 and 3/2016. Is it march 2016 (and a little late) or Q3 2016?

**Comment by haskida** [comment]
@chrigrahcisco 
Can you provide more detail on the 1.5.X server fix?
I am using C# client with 1.5.3 server and IOSDriver (not the new beta driver).
I and still fail to Driver.SwitchTo().Alert().Accept(); from time to time.

**Comment by chrigrahcisco** [comment]
The only thing that's changed from 1.4.x to 1.5.x is that Appium no longer fails to find my AUT due to the Notification system alert. I no longer am required to inject a sleep into the AUT source code as described in detail above.

**Comment by haskida** [comment]
@chrigrahcisco 
I guess my next question is, how do yo accept the alerts?
Just a simple switchTo.alert.accept?

Thanks for the prompt reply

**Comment by chrigrahcisco** [comment]
Correct, but wrapped a loop with a 10 second timeout. This was written way back in Appium 1.4.x along with injecting the sleep into the source code.  

Based on what I've seen with Appium 1.5.x I'm guessing the loop and timeout are not even needed any longer, but I haven't bothered to find out since everything is working without error.

This is called in our setup function for every testcase.  Python 2.7.x

`def dismiss_startup_notification(self):
    try:
        # Utilizes the native Alert handling handled directly by the driver
        wait_time = 10
        print "Waiting for Startup notification dialog to open, max wait =", wait_time, "seconds"
        timeout = time.time() + wait_time   # Timer based on wait_time to prevent infinite loops
        while True:
            alert = self.driver.switch_to.alert   # If the dialog open
            if "Send You Notifications" in alert.text:
                    print "Found startup notification alert"
                    alert.dismiss()
                    print "Dismissed startup notification alert"
                    return True
            if time.time() > timeout:
                print "Warning! Startup notification dialog still NOT open after", wait_time, "seconds"
                break
            print "Sleeping 1 second"
            time.sleep(1)   # Prevent CPU slamming with short timeout between loops
    except:
        print "Startup notification dialog not found or failed to close."
        return False`

Hope this helps.

**Comment by pengboZheng** [comment]
Appium: 1.5.3
iOS: 9.3.5
Only delay 3 seconds before register the notification, the tests are pretty stable now. Thank you @chrigrahcisco for the tips!

**Comment by jlipps** [comment]
Closing this issue since I don't think there's anything we can do on the Appium side to fix this, plus Instruments+UIAutomation is no more, so hopefully this issue will not plague us in the XCUITest world.

**Comment by sandhya1990** [comment]
I have used 
driver.switchTo()a.ert().accept() in my code

it accepts the notification alert pop-up and other permission pop-ups but app crashes as soon as the alerts are closed.

I am running app file on simulator .

Looking forward for the solution.

**Comment by sandhya1990** [comment]
One more issue, even after accepting notification alert,  during login an alert pop-pop's up as enable notiifcation
